src/combinatorics/derangements/basic.lean

```diff
@@ -0,0 +1,190 @@
+fixed iff they are not in the subtype. -/
+protected def subtype_equiv (p : α → Prop) [decidable_pred p] :
+  derangements (subtype p) ≃ {f : perm α // ∀ a, ¬p a ↔ a ∈ fixed_points f} :=
+calc
+  derangements (subtype p)
+      ≃ {f : {f : perm α // ∀ a, ¬p a → a ∈ fixed_points f} // ∀ a, a ∈ fixed_points f → ¬p a}
+      : begin
+        refine (perm.subtype_equiv_subtype_perm p).subtype_equiv (λ f, ⟨λ hf a hfa ha, _, _⟩),
+        { refine hf ⟨a, ha⟩ (subtype.ext _),
+          rw [mem_fixed_points, is_fixed_pt, perm.subtype_equiv_subtype_perm] at hfa,
+          dsimp at hfa,
+          rwa equiv.perm.of_subtype_apply_of_mem at hfa },
+        rintro hf ⟨a, ha⟩ hfa,
+        refine hf _ _ ha,
+        change perm.subtype_equiv_subtype_perm p f a = a,
+        rw [perm.subtype_equiv_subtype_perm_apply_of_mem f ha, hfa, subtype.coe_mk],
+      end
+  ... ≃ {f : perm α // ∃ (h : ∀ a, ¬p a → a ∈ fixed_points f), ∀ a, a ∈ fixed_points f → ¬p a}
+      : subtype_subtype_equiv_subtype_exists _ _
+  ... ≃ {f : perm α // ∀ a, ¬p a ↔ a ∈ fixed_points f}
+      : subtype_equiv_right (λ f, by simp_rw [exists_prop, ←forall_and_distrib,
+        ←iff_iff_implies_and_implies])
+
+/-- The set of permutations that fix at either `a` or nothing is equivalent to the sum of:
+    - derangements on `α`
+    - derangements on `α` minus `a`. -/
+def at_most_one_fixed_point_equiv_sum_derangements [decidable_eq α] (a : α) :
+  {f : perm α // fixed_points f ⊆ {a}} ≃ (derangements ({a}ᶜ : set α)) ⊕ derangements α :=
+calc
+  {f : perm α // fixed_points f ⊆ {a}}
+      ≃ {f : {f : perm α // fixed_points f ⊆ {a}} // a ∈ fixed_points f}
+        ⊕ {f : {f : perm α // fixed_points f ⊆ {a}} // a ∉ fixed_points f}
+      : (equiv.sum_compl _).symm
+  ... ≃ {f : perm α // fixed_points f ⊆ {a} ∧ a ∈ fixed_points f}
+        ⊕ {f : perm α // fixed_points f ⊆ {a} ∧ a ∉ fixed_points f}
+      : begin
+        refine equiv.sum_congr _ _;
+        { convert subtype_subtype_equiv_subtype_inter _ _, ext f, refl }
+      end
+  ... ≃ {f : perm α // fixed_points f = {a}} ⊕ {f : perm α // fixed_points f = ∅}
+      : begin
+        refine equiv.sum_congr (subtype_equiv_right $ λ f, _) (subtype_equiv_right $ λ f, _),
+        { rw [set.eq_singleton_iff_unique_mem, and_comm],
+          refl },
+        { rw set.eq_empty_iff_forall_not_mem,
+          refine ⟨λ h x hx, h.2 (h.1 hx ▸ hx), λ h, ⟨λ x hx, (h _ hx).elim, h _⟩⟩ }
+      end
+  ... ≃ (derangements _) ⊕ derangements α
+      : begin
+        refine equiv.sum_congr ((derangements.subtype_equiv _).trans $ subtype_equiv_right $ λ x,
+          _).symm (subtype_equiv_right $ λ f, mem_derangements_iff_fixed_points_eq_empty.symm),
+        rw [eq_comm, set.ext_iff],
+        simp_rw [set.mem_compl_iff, not_not],
+      end
+
+namespace equiv
+variables [decidable_eq α]
+
+/-- The set of permutations `f` such that the preimage of `(a, f)` under
```

Comment:
```suggestion
/-- The set of permutations `f` such that a preimage of `(a, f)` under
```
or is it really unique?

Reply:
it's unique once `a` is fixed, since `equiv.perm.decompose_option` is an equivalence